lib/iris/coords.py

```diff
@@ -2775,6 +2778,469 @@ def xml_element(self, doc):
+        if indices_min < self.start_index:
+            indices_error(
+                f"Lowest index: {indices_min} < start_index: {self.start_index} ."
+            )
+
+        indices_shape = indices.shape
+        if len(indices_shape) != 2:
+            indices_error(
+                f"Expected 2-dimensional shape, got: shape={indices_shape} ."
+            )
+
+        len_req_fail = False
+        if shapes_only:
+            src_shape = indices_shape[self.tgt_dim]
+            # Wrap as lazy to allow use of the same operations below
+            # regardless of shapes_only.
+            src_lengths = _lazy.as_lazy_data(np.asarray(src_shape))
+        else:
+            src_lengths = self._lazy_src_lengths(indices)
+        if self.src_location in ("edge", "boundary"):
+            if (src_lengths != 2).any().compute():
+                len_req_fail = "len=2"
+        else:
+            if self.src_location == "face":
+                min_size = 3
+            elif self.src_location == "volume":
+                if self.tgt_location == "edge":
+                    min_size = 6
+                else:
+                    min_size = 4
+            else:
+                raise NotImplementedError
+            if (src_lengths < min_size).any().compute():
+                len_req_fail = f"len>={min_size}"
+        if len_req_fail:
+            indices_error(
+                f"Not all src_locations meet requirement: {len_req_fail} - "
+                f"needed to describe '{self.cf_role}' ."
+            )
+
+    def validate_indices(self):
+        """
+        Perform a thorough validity check of this connectivity's
+        :attr:`indices`. Includes checking the sizes of individual
+        :attr:`src_location`'s (specified using masks on the
+        :attr:`indices` array) against the :attr:`cf_role`.
+
+        Raises a ``ValueError`` if any problems are encountered, otherwise
+        passes silently.
+
+        .. note::
+
+            While this uses lazy computation, it will still be a high
+            resource demand for a large :attr:`indices` array.
+
+        """
+        self._validate_indices(self.indices, shapes_only=False)
+
+    def __eq__(self, other):
+        eq = NotImplemented
+        if isinstance(other, Connectivity):
+            # Account for the fact that other could be the transposed equivalent
+            # of self, which we consider 'safe' since the recommended
+            # interaction with the indices array is via indices_by_src, which
+            # corrects for this difference. (To enable this, src_dim does
+            # not participate in ConnectivityMetadata to ConnectivityMetadata
+            # equivalence).
+            if hasattr(other, "metadata"):
+                # metadata comparison
+                eq = self.metadata == other.metadata
+                eq = (
+                    eq
+                    and (self.indices_by_src() == other.indices_by_src()).all()
+                )
```

Comment:
@trexfeathers This could be simplified to the following pattern:
```python
if eq:
    eq = (self.indices_by_src() == other.indices_by_src()).all()
```

Reply:
6c9c76510b80fe3f9fae7a81ac1a449b0a577855